### PR TITLE
[table] Improved documentation

### PR DESCRIPTION
**Pull request by tmorehouse**
Added more detail about scoped slots and  supported `item` and `field` properties

### CODE REVIEW

**Comment by alexsasharegan** [comment]
Is the `fields` prop supposed to be an object?

File: docs/components/table/README.md [hunk 1]
```diff
@@ -17,28 +17,107 @@ age: {
...
+| state | String | Bootstrap contextual state applied to row (`active`, `success`, `info`, `warning`, `danger`)
+
+
 ### Custom rendering
 Custom rendering for each field is possible using **scoped slots**.
 If you want to add an extra field which does not exits on records, just add it to `fields` array.  Example:
- 
+
+```js
+ fields: [
```
Comment (alexsasharegan) [inline docs/components/table/README.md]: Are the fields supposed to be an array of objects, or just an object?

**Comment by tmorehouse** [comment]
Oops, yeah.  I'll fix that

File: docs/components/table/README.md (hunk 1 above)
Comment (tmorehouse) [inline docs/components/table/README.md]: Oops, yeah. I'll fix that.  Thanks for catching that! :+1:

Comment (alexsasharegan) [inline docs/components/table/README.md]: Days since last Troy-o: `0`

**Comment by tmorehouse** [comment]
I'm wondering if the item prop `state` should be `_state` to differentiate it from record values?

We could make `state` deprecated and start to use `_state`.

Also, it appears the default `toString` method for filtering/searching includes the `state` variable.  We probably should filter that out of of the item before converting it to a string for filtering.

As well, filtering only looks at the raw data, and doesn't include any virtual computed columns.  This might need to be noted in the documentation.

**Comment by alexsasharegan** [comment]
Hmm... I was looking into what you mentioned, and I came across something else. Does the default function for the prop `field` actually return an object? Arrow function shorthand to return an object literal is ambiguous due to the code block syntax. Take a look here: https://github.com/bootstrap-vue/bootstrap-vue/blob/master/lib/components/table.vue#L96-L97

image
image

**Comment by alexsasharegan** [comment]
I'm kind of a constructor hater for Arrays and Objects, but it's probably better than this: 

image

**Comment by alexsasharegan** [comment]
So I can see how state might collide with a user value. I'm not a fan of using private notation, but how about `bState` and the html equivalent `b-state`?

Also, I don't see how the toString method incorporates the `state` since it uses that `sortBy` variable to pluck of the appropriate item field.

Good call on the itemProvider. Looking through the source was the first I saw of it.

**Comment by alexsasharegan** [comment]
P.S. `key` on line 249 is undefined

**Comment by tmorehouse** [comment]
I'll do  quick commit to add key to the fieldClass method

**Comment by tmorehouse** [comment]
fieldClass is now fixed with comit https://github.com/bootstrap-vue/bootstrap-vue/commit/2900e953a49b1941be8481e1bac89cf646249666

**Comment by tmorehouse** [comment]
@alexsasharegan The filter function converts the item row object to a string at line 220 https://github.com/bootstrap-vue/bootstrap-vue/blob/master/lib/components/table.vue#L220

So anything in the item object gets converted to a string representation (i.e. including the `state` property (if it exists).

**Comment by alexsasharegan** [comment]
Ah, I see that now. Is there a way to confine that to the value prop name?

**Comment by tmorehouse** [comment]
About the default for fields, yeah, the `{}` is ambiguous in `default: () => {}`

It might be better to make it better/clearer if it was:

 `default: () => { return {}; }`

 or

 `default: () => new Object()`

**Comment by tmorehouse** [comment]
We could use the fields object keys to select the item keys to stringify (i.e. by making an intermediate object to stringify)

**Comment by tmorehouse** [comment]
It mght also be nice to allow computed fields in the items array (i.e. an anonymous function), but that might be too much bloat on the items array (and the user could run their own massage routine on the items using the `itemsProvider` function.

**Comment by tmorehouse** [comment]
also, trying to figure out what the prop `item_key` is for.  It is bound to each `tr` in the table body.  And it doesn't follow normal camelCase / kebab-csae naming.

**Comment by alexsasharegan** [comment]
Lots of good ideas! And you've got your hands full with this one 😃 

Is this a good time to bring up that it's kinda weird to have the `<tfoot>` above the `<tbody>`?

**Comment by tmorehouse** [comment]
Yeah, I used to always poot my `<tfoot>` at the bottom of the table, but based on the W3 html5 standards the `<tfoot>` follows directly below the `<thead>`.

I guess this is because, in some printing situation, the `<tfoot>` (and `<thead>`) can appear on multiple pages if the table is longer than a page. (although I haven't stumbled across any browser that does that).

**Comment by alexsasharegan** [comment]
Welp, learn something new every day!

**Comment by tmorehouse** [comment]
And regarding itemsProvider, it might be handy to add in a `Boolean` prop to signal that paging is done in the `itemProvider` function. maybe `providerPaging`.  This would allow `b-table` to maintain the state of paging, but not use it when creating `_items`.  As well, filtering could be done via the provider function as well (for server side filtering) if another `Boolean` prop. is set (say `providerFilter`).

A watch method could see if any of the paging or filter values have changed and then freshen the `items/_items` array.

Could address issues like #394 and #131

**Comment by tmorehouse** [comment]
@pi0 & @alexsasharegan 

Just thinking, `state` prop on item is really a `variant` class for the `tr`.

maybe `state` could be set as `_variant` (semi-private item prop) to follow with the other places we use `variant` as a component prop.

`state` could be deprecated fro preference of the `_variant` prop:

```html
<tr v-for="(item,index) in _items"
    role="row"
    :key="items_key"
    :class="[(item._variant || item.state)?'table-'+(item._variant || item.state) : null]"
    @click="rowClicked(item, index)"
>
```

And then a few releases later we could remove `item.state` support.

The row's class could be made into a method for computing the class as well as `rowClass(item)`

**Comment by tmorehouse** [comment]
Or make the `state` property `_rowVariant`, which better reflects that is a row level property

**Comment by alexsasharegan** [comment]
Good catch. I think we should attempt to unify the API around using `variant`. I'm gonna vote for using plain `rowVariant` though just because I think it's super awkward having a public API prop use that notation. One of those _just doesn't feel right_ type things to me. With the extra specificity of the `row` prefix, I think does a good job of both avoiding collisions and improving semantics.

**Comment by tmorehouse** [comment]
To make sure it is not a breaking change, I could use the logic above to prefer `_rowVariant` over `state`.

**Pull request merged**